src/tree/builder/TreeAdapter.js

```diff
@@ -0,0 +1,449 @@
+// HTML classes.
+import { formattingElements, headings, irrelevantHtmlElements } from "./htmlClasses";
+
+/**
+ * An adapter to parse the HTML source code to a structured tree representation, to be used in further analysis,
+ * in combination with the `parse5` library.
+ *
+ * Implements the `parse5` library's `TreeAdapter` interface
+ * @see https://github.com/inikulin/parse5/blob/master/packages/parse5/docs/tree-adapter/interface.md
+ *
+ * @memberOf module:tree/builder
+ *
+ * @private
+ */
+class TreeAdapter {
+	// Creation of nodes and other tree elements.
+
+	/**
+	 * Creates a new element to be put in the tree.
+	 *
+	 * @param {string} tag									The HTML tag.
+	 * @param {string} namespace							The XML namespace (e.g. "http://www.w3.org/1999/xhtml" for HTML).
+	 * @param {Array<{ name: string, value: string }>} attributes	The attributes of this element.
+	 *
+	 * @returns {module:tree/structure.Node|module:tree/structure.FormattingElement} The new element.
+	 */
+	createElement( tag, namespace, attributes ) {
+		let node;
+
+		if ( irrelevantHtmlElements.includes( tag ) ) {
+			// Irrelevant for analysis (e.g. `script`, `style`).
+			node = new StructuredIrrelevant( tag );
+		} else if ( formattingElements.includes( tag ) ) {
+			// Formatting element.
+			const parsedAttributes = TreeAdapter._parseAttributes( attributes );
+			node = new FormattingElement( tag, parsedAttributes );
+		} else {
+			// Paragraphs, Headers, Lists, ListItems and other nodes.
+			node = this._parseNode( tag );
+		}
+
+		node.tagName = tag;
+		node.namespace = namespace;
+		node.parent = null;
+
+		return node;
+	}
+
+	/**
+	 * Makes a new node to add to the tree, based on the given HTML-tag.
+	 *
+	 * @param {string} tag		The HTML-tag of the element to add.
+	 *
+	 * @returns {module:tree/structure.Node} The node to add to the tree.
+	 *
+	 * @private
+	 */
+	_parseNode( tag ) {
+		if ( headings.includes( tag ) ) {
+			// Heading.
+			return new Heading( parseInt( tag[ 1 ], 10 ) );
+		} else if ( tag === "p" ) {
+			// Paragraph.
+			return new Paragraph( tag );
+		} else if ( tag === "ol" || tag === "ul" ) {
+			// List.
+			return new List( tag === "ol" );
+		} else if ( tag === "li" ) {
+			// List item.
+			return new ListItem();
+		}
+		// All other elements (`div`, `section`).
+		return new StructuredNode( tag );
+	}
+
+	/**
+	 * Parses the HTML element attributes from parse5's format to a plain JS object.
+	 * E.g. `{ name: "id", value: "an-id" }` becomes ` { id: "an-id" }`.
+	 *
+	 * @param {Array<{ name: string, value: string }>} parse5attributes		The attributes as parsed by parse5.
+	 *
+	 * @returns {Object} The parsed attributes.
+	 *
+	 * @private
+	 */
+	static _parseAttributes( parse5attributes ) {
+		if ( parse5attributes ) {
+			return parse5attributes.reduce( ( attributes, attribute ) => {
+				attributes[ attribute.name ] = attribute.value;
+				return attributes;
+			}, {} );
+		}
+		return {};
```

Comment:
I changed it, it now returns `null` instead of the empty object. In general: the Wiki on the tree structure is outdated, I will update the Wiki after the code has stabilized a bit, to avoid superfluous work.

I added an extra test for the `TreeAdapter` to check if the attributes are `null` when a formatting element with no attributes is added.